### PR TITLE
8292226: Prepare make for better Link Time Optimization support

### PR DESCRIPTION
**Pull request by TheShermanTanker**
The support for Link Time Optimization in the JDK's make system could do with some cleaning up, at the moment it simply assumes the compiler is gcc and sets the flags as such. Instead of introducing changes in bulk, as a first step, it would be good to simply supply the appropriate flags depending on the compiler and refine the flags for the one existing compiler with support for it. In practice the latter just means adding the proper -fuse-linker-plugin to gcc compile step when link-time-opt is specified, as without it the compiler generates both native code and information rich representation for Link Time Optimization. With this flag, native code generation is disabled and object files will contain only code required for Link Time Optimization, thus speeding up compile times as well.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292226](https://bugs.openjdk.org/browse/JDK-8292226): Prepare make for better Link Time Optimization support

### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9829/head:pull/9829` \
`$ git checkout pull/9829`

Update a local copy of the PR: \
`$ git checkout pull/9829` \
`$ git pull https://git.openjdk.org/jdk pull/9829/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9829`

View PR using the GUI difftool: \
`$ git pr show -t 9829`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9829.diff">https://git.openjdk.org/jdk/pull/9829.diff</a>

</details>

### CODE REVIEW

File: make/hotspot/lib/JvmFeatures.gmk [hunk 1]
```diff
@@ -169,12 +169,14 @@ endif
...
 
 ifeq ($(call check-jvm-feature, link-time-opt), true)
-  # NOTE: Disable automatic opimization level and let the explicit cflag control
-  # optimization level instead. This activates O3 on slowdebug builds, just
-  # like the old build, but it's probably not right.
-  JVM_OPTIMIZATION :=
-  JVM_CFLAGS_FEATURES += -O3 -flto
-  JVM_LDFLAGS_FEATURES += -O3 -flto -fuse-linker-plugin -fno-strict-aliasing
+  ifeq ($(call isCompiler, gcc), true)
+    # NOTE: Disable automatic opimization level and let the explicit cflag control
```
Comment (dholmes-ora) [inline make/hotspot/lib/JvmFeatures.gmk:173]: existing typo: opimization

**Review by dholmes-ora — COMMENTED** [review]
I'll leave it to build folk to approve this. Making this contingent on using gcc seems reasonable. But I think we may end up just removing this anyway.

Thanks.

File: make/hotspot/lib/JvmFeatures.gmk (hunk 1 above)
Comment (TheShermanTanker) [reply make/hotspot/lib/JvmFeatures.gmk]: Fixed it, thanks for catching it

**Comment by TheShermanTanker** [comment]
> But I think we may end up just removing this anyway.

You never know for certain if this might come in handy in the future ;)

**Comment by magicus** [comment]
Technically this change seems reasonable, but I tend to agree with David that it is more likely that a better solution would be to remove this completely. Have you tried building with LTO enabled? What kind of build times do you get -- and what kind of improvements in terms of file size and performance? My gut feeling is that there is no way that the small potential improvements can counteract the huge negative impact of build times magnitudes larger.

**Review by magicus — APPROVED** [review]
Whatever. Your patch makes the code less ugly, and possibly LTO a bit more useful. Let's save the "shall we remove this?" discussion for another day.

**Pull request closed without merging**